File: lib/mnesia_session/examples/person/erl_session/person_ex.erl
%% ``The contents of this file are subject to the Erlang Public License,
%% Version 1.1, (the "License"); you may not use this file except in
%% compliance with the License. You should have received a copy of the
%% Erlang Public License along with this software. If not, it can be
%% retrieved via the world wide web at http://www.erlang.org/.
%% 
%% Software distributed under the License is distributed on an "AS IS"
%% basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See
%% the License for the specific language governing rights and limitations
%% under the License.
%% 
%% The Initial Developer of the Original Code is Ericsson Utvecklings AB.
%% Portions created by Ericsson are Copyright 1999, Ericsson Utvecklings
%% AB. All Rights Reserved.''
%% 
%%     $Id$
%%
%%%----------------------------------------------------------------------
%%% Purpose : Examples of mnesia_sesion interface to mnesia
%%%----------------------------------------------------------------------

-module(person_ex).

-compile(export_all).
%%-export([Function/Arity, ...]).

%% Use setup:go() to initiate the example

%%%Includes
%% Include to get the mnesia types, used in create table
-include_lib("mnesia_session/include/mnesia.hrl").   

%% Include my own definitions
-include("persons.hrl").  
%%%Includes

run() ->
    run(node()).

run(ServerNode) ->
    {ConnPid, SessPid} = start_session(ServerNode),
    io:format("Create a persons_person table~n"),
    create_person_table(SessPid),
    PD = insert_person(SessPid, "Dan", male, 27, 97185, "none", []),    
    io:format("Inserted person ~p in mnesia~n", [PD]),
    PN = insert_person(SessPid, "Niklas", male, 38, 99999, "Lisa", ["Jonas", "Ola"]), 
    io:format("Inserted person ~p in mnesia~n", [PN]),
    RES = get_person(SessPid, "Dan"),
    io:format("Read person ~p from mnesia ~p ~n", ["Dan", RES]),
    mnesia_connector:disconnect(ConnPid, SessPid).

% Starts a mnesia_session object
%%%Init
start_session(Node) ->
    ConnPid = rpc:call(Node, mnesia_session_lib, lookup_connector, []),
    %% Use the factory to create a session
    SessPid = mnesia_connector:connect(ConnPid),
    {ConnPid, SessPid}.
%%%Init

%% Some examples of how to use the mnesia_session interface!

%%%create_table
create_person_table(ObjKey) ->
    %% Define the table properties
    Attrs = [atom_to_list(F) || F <- record_info(fields, persons_person)],
    TabDef = #mnesia_TableDef{type = bag, mode = read_write, 
			      ram_copies = [],
			      disc_copies = [], 
			      disc_only_copies = [],
			      index_list = [4],  %% Index on married_to
			      attributes = Attrs,
			      %% OBSERVE that the record_name must be 
			      %% exactly the same as the name of the 
			      %% structure/record  
			      record_name = "persons_person"},
    
    Res = mnesia_session:create_table(ObjKey, "persons", TabDef),
    case Res of 
	{ok, ""} ->	   
	    ok;
	Else->
	    io:format("ERROR: ~s~n", [Else])
    end.
%%%create_table

%%%dirty_write
insert_person(SessionKey, Name, Sex, Age, Phone, Mt, Ch) 
  when list(Name), atom(Sex), integer(Age), 
       integer(Phone), list(Mt), list(Ch) ->
    
    Data = #persons_data{sex = Sex, age = Age, phone = Phone},
    Person = #persons_person{name = Name, personData = Data, 
			     married_to = Mt, children = Ch},
    
    {ok, ""} = mnesia_session:dirty_write(SessionKey, "persons", Person),
    Person.
%%%dirty_write

%%%dirty_read
get_person(SessionKey, Name) when list(Name) ->
    {ok, RObj, ""} = 
	mnesia_session:dirty_read(SessionKey, "persons", Name),    
    hd(RObj).
%%%dirty_read

File: lib/mnesia_session/examples/person/java_session/person_ex.java

///Imports
import com.ericsson.otp.ic.*;
///Imports

public class person_ex
{
    public static void main(String args[])
    {
      mnesia._sessionStub mcsRef = null;
      
      mcsRef = start_session(args);
      if(mcsRef != null)
	{
	  sysinfo(mcsRef);
	  create_person_table(mcsRef);
	  insert_person(mcsRef, "Dan", persons.Sex.male, 
			27, 97185, "none", new String[0]);
	  System.out.println("Wrote person Dan ");
	  persons.person p;
	  
	  p = get_person(mcsRef, "Dan");
	  if(p != null)
	    {
	      System.out.println("Read person Dan from mnesia");
	      System.out.println(" Sex " + p.personData.sex.value() + 
				 " Age " +  p.personData.age +
				 " Phone " + p.personData.phone);
	    }
	}
      else
	System.out.println("Ok this didn't work try something else!");
    }

///create_table
  public static void create_person_table(mnesia._sessionStub mcsRef)
    {
      try {
      
	String name = "persons";
	mnesia.TableDef def = new mnesia.TableDef();
	def.type = mnesia.SetOrBag.bag;
	def.mode = mnesia.AccessMode.read_write;
	def.ram_copies = new String[0];
	def.disc_copies = new String[0];
	def.disc_only_copies = new String[0];
	int[] idxs = new int[1];
	idxs[0] = 4;
	def.index_list = idxs;
	String[] attrs = new String[4];
	attrs[0] = "name";
	attrs[1] = "personData";
	attrs[2] = "married_to";
	attrs[3] = "children";
	def.attributes = attrs;
	def.record_name = "persons_person"; // The used IDL type
	
	StringHolder reason;
	reason = new StringHolder();
	
	if(mnesia.Status.ok != mcsRef.create_table(name, def, reason))
	  System.out.println("Create Table Error " + reason.value); 
      }
      catch(Exception se)
	{
	  System.out.println("Unexpected exception: " + se.toString());
	  return;
	} 
    }
///create_table

///dirty_write
  public static void insert_person(mnesia._sessionStub mcsRef,
				   String name,
				   persons.Sex sex,
				   int age,
				   int phone,
				   String mt,
				   java.lang.String[] children)
    {
      persons.data data;
      data = new persons.data(sex, age, phone);
      persons.person person = new persons.person();
      person.name = name;
      person.personData = data;
      person.married_to = mt;
      person.children = children;
      
      try
	{
	  StringHolder reason = new StringHolder();       
	  Term object = new Term();
	  persons.personHelper.insert(object,person);
	  
	  if(mnesia.Status.ok != mcsRef.dirty_write("persons", object, reason))
	    System.out.println("Insert person Error " + reason.value);
	}
      catch(Exception se)
	{
	  System.out.println("Unexpected exception: " + se.toString());
	  return;
	}
    }
///dirty_write
  
///dirty_read
  public static persons.person 
    get_person(mnesia._sessionStub mcsRef, String name)
    {	
      try
	{
	  StringHolder reason = new StringHolder();       
	  Term key = new Term();
	  mnesia.RecordlistHolder res = new mnesia.RecordlistHolder();
	  key.insert_string(name);

	  if(mnesia.Status.ok == mcsRef.dirty_read("persons", 
						   key, res, reason))
	    {
	      if(res.value.length > 0) 
		{
		  persons.person rec1 = 
		    persons.personHelper.extract(res.value[0]);

		  return rec1;
		}
	      else
		return null;			    
	    }
	  else
	    {
	      System.out.println("Insert person Error " + reason.value);
	      return null;
	    }
	}
	catch(Exception se)
	  {
	    System.out.println("Unexpected exception: " + se.toString());
	    return null;
	  }	
    }
///dirty_read
  
///sysinfo
  public static void sysinfo(mnesia._sessionStub mcsRef)
    { 
      StringHolder reason;
      reason = new StringHolder();
	
      mnesia.SystemInfoHolder info = new mnesia.SystemInfoHolder();
      try
	{	    		
	  if(mnesia.Status.ok == mcsRef.system_info(info, reason))
	    {
	      System.out.println("------------------------------------------------");
	      System.out.println("System Info ");
	      System.out.println("Mnesia is_running? " + info.value.is_running);
	      System.out.println("backup_module: " + info.value.backup_module);
	      System.out.println("Directory: " + info.value.directory);
	      System.out.println("Dump Log time time threshold " + 
				   info.value.dump_log_time_threshold + " ms.");
	      System.out.println("------------------------------------------------");
	    }
	  else
	    System.out.println("Info Error" + reason.value);
	}
      catch(Exception se)
	{
	  System.out.println("Unexpected exception: " + se.toString());
	  return;
	}  
    }
///sysinfo

///Init
  public static mnesia._sessionStub start_session(String args[])
    {
      String SNode = new String(args[0]);
      String PNode = new String(args[1]);
      String Cookie = new String(args[2]);

      mnesia._connectorStub mccRef = null;
      mnesia._sessionStub mcsRef = null;
	       
      try
	{
	  
	  mccRef = new mnesia._connectorStub(SNode+"0",PNode,Cookie,
					     "mnesia_connector");
	  
	  com.ericsson.otp.ic.Pid mccPid = mccRef.connect();
	  
	  mcsRef = new mnesia._sessionStub(SNode+"1",PNode,Cookie,mccPid);

	  return mcsRef;
	}
      catch(Exception se)
	{
	  System.out.println("Unexpected exception: " + se.toString());
	  se.printStackTrace();
	  return null;
	}
    }
///Init

}




